mantidproject/mantid
# Workbench sliceviewer: Dynamically rebin MDEventWorkspaces on zoom

**Description of work.**

Adds dynamic rebinning to the sliceviewer. When the axes limits change on the view, if the workspace is an MDEventWorkspace, then the data is rebinned to the limits of the view rather than the full extent of the dimensions.

Zooming on scroll of the mouse wheel has also been enabled.

**To test:**

The following script can be used for a basic test of the functionality. It creates an `MDEventWorkspace`, fakes some data and creates some fake peaks.

Run the script and show the sliceviewer and play with the zooming:
* zoom with the magnifier tool and reset up to home - ensure the limits are as expected when drawing the zoom rectangle after the zooming completes
* zoom on scroll and check the data is as expected
* activate the peaks viewer and ensure the peak selection still zooms to the peak and when zooming in on the data the peaks stays where it should.
* check the non-orthogonal view works and vary the setub parameters to check

```python
mdws = CreateMDWorkspace(Dimensions=3,
                         Extents='-10,10,-10,10,-10,10',
                         Names='A,B,C',
                         Units='r.l.u.,r.l.u.,r.l.u.',
                         Frames='HKL,HKL,HKL')
expt_info = CreateSampleWorkspace()
mdws.addExperimentInfo(expt_info)
a, b, c = 1, 1, 1
alpha, beta, gamma = 90, 90, 90
SetUB(mdws, a, b, c, alpha, beta, gamma)
FakeMDEventData(InputWorkspace=mdws, PeakParams='500000,0,0,0,3')
run_number = 1
sxd_ws = LoadEmptyInstrument(InstrumentName='SXD')
AddSampleLog(sxd_ws, 'run_number', str(run_number), 'Number')
peak_ws = CreatePeaksWorkspace(InstrumentWorkspace=sxd_ws, NumberOfPeaks=0)
peak_ws.mutableRun().addProperty('run_number', run_number, True)
SetUB(peak_ws, a, b, c, alpha, beta, gamma)
# Add a new peak
AddPeakHKL(peak_ws, [2, 0, 0])
AddPeakHKL(peak_ws, [2, 1, 4])
# integrate to get a sphere with a background
peak_ws_sphere = IntegratePeaksMD(InputWorkspace=mdws,
                                  PeaksWorkspace=peak_ws, PeakRadius=0.2,
                                  BackgroundInnerRadius=0.7,
                                  BackgroundOuterRadius=0.9)

```

Also check that opening the viewer on the `mdws_svrebinned` actually shows the view for the original `mdws` but updates the window title to show this.

Pull request #28506 also has a script and data links to real examples. Test these too.

Fixes #28689
Fixes #28579 

*This does not require release notes* because **we'll update the sliceviewer release notes in one go before 5.1**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
